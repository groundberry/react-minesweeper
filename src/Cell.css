.Cell {
  height: 25px;
  width: 25px;
}

.Cell--button {
  background-color: #ccc;
  border-color: #d8d8d8;
  border-style: outset;
  border-width: medium;
  color: #000;
  text-align: center;
}

.Cell--button:hover {
  background-color: #dcdcdc;
  border-color: #d0d0d0;
  border-style: outset;
  border-width: medium;
  cursor: pointer;
}

.Cell--image {
  background-color: #ccc;
  background-size: 100% 100%;
}
