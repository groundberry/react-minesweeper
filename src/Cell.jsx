import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { getCellImage } from './utils/cell';
import './Cell.css';

class Cell extends Component {
  constructor() {
    super();

    this.state = {
      flagged: false,
    };

    this.handleClickShowContent = this.handleClickShowContent.bind(this);
    this.handleClickShowFlag = this.handleClickShowFlag.bind(this);
  }

  handleClickShowContent() {
    const { flagged } = this.state;
    const {
      coordinates,
      content,
      onClickCell,
      onClickMine,
    } = this.props;

    if (flagged) {
      return;
    }

    if (content === '*') {
      onClickMine();
    }

    onClickCell(coordinates);
  }

  handleClickShowFlag(evt) {
    evt.preventDefault();

    this.setState(prevState => ({ flagged: !prevState.flagged }));
  }

  render() {
    const { flagged } = this.state;
    const { content, showContent } = this.props;
    const pressedImg = getCellImage(content);
    const flagImg = flagged && getCellImage('f');

    return (
      <Fragment>
        {showContent ? (
          <div
            style={{ backgroundImage: `url(${pressedImg})` }}
            className="Cell Cell--image"
          />
        ) : (
          <button
            style={{ backgroundImage: `url(${flagImg})` }}
            className="Cell Cell--button Cell--image"
            onClick={this.handleClickShowContent}
            onContextMenu={this.handleClickShowFlag}
          />
        )}
      </Fragment>
    );
  }
}

Cell.propTypes = {
  coordinates: PropTypes.arrayOf(PropTypes.number).isRequired,
  content: PropTypes.string.isRequired,
  showContent: PropTypes.bool.isRequired,
  onClickCell: PropTypes.func.isRequired,
  onClickMine: PropTypes.func.isRequired,
};

export default Cell;
