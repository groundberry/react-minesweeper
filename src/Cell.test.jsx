import React from 'react';
import { shallow } from 'enzyme';
import Cell from './Cell';

describe('Cell', () => {
  it('renders one cell component', () => {
    const onClickMine = jest.fn();
    const onClickCell = jest.fn();
    const coordinates = [0, 1];
    const showContent = false;

    const wrapper = shallow(<Cell
      content="1"
      coordinates={coordinates}
      onClickMine={onClickMine}
      onClickCell={onClickCell}
      showContent={showContent}
    />);
    expect(wrapper).toMatchSnapshot();
  });
});
