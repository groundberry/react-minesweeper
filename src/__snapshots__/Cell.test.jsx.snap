// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell when the game is not over with a normal cell renders one cell component 1`] = `
<React.Fragment>
  <button
    className="Cell Cell--button Cell--image"
    onClick={[Function]}
    onContextMenu={[Function]}
    style={
      Object {
        "backgroundImage": "url(false)",
      }
    }
  />
</React.Fragment>
`;
